qiskit/compiler/transpile.py

```diff
@@ -308,8 +313,68 @@ def _transpile_circuit(circuit_config_tuple: Tuple[QuantumCircuit, Dict]) -> Qua
         pass_manager.append(MSBasisDecomposer(ms_basis_swap))
 
-    return pass_manager.run(circuit, callback=transpile_config['callback'],
-                            output_name=transpile_config['output_name'])
+    result = pass_manager.run(circuit, callback=transpile_config['callback'],
+                              output_name=transpile_config['output_name'])
+
+    if transpile_config['faulty_qubits_map']:
+        return _remap_circuit_faulty_backend(result, transpile_config['backend_num_qubits'],
+                                             pass_manager_config.backend_properties,
+                                             transpile_config['faulty_qubits_map'])
+
+    return result
+
+
+def _remap_circuit_faulty_backend(circuit, num_qubits, backend_prop, faulty_qubits_map):
+    faulty_qubits = backend_prop.faulty_qubits() if backend_prop else []
+    disconnected_qubits = {k for k, v in faulty_qubits_map.items()
+                           if v is None}.difference(faulty_qubits)
+    faulty_qubits_map_reverse = {v: k for k, v in faulty_qubits_map.items()}
+    if faulty_qubits:
+        faulty_qreg = circuit._create_qreg(len(faulty_qubits), 'faulty')
+    else:
+        faulty_qreg = []
+    if disconnected_qubits:
+        disconnected_qreg = circuit._create_qreg(len(disconnected_qubits), 'disconnected')
+    else:
+        disconnected_qreg = []
+
+    new_layout = Layout()
+    faulty_qubit = 0
+    disconnected_qubit = 0
+
+    for real_qubit in range(num_qubits):
+        if faulty_qubits_map[real_qubit] is not None:
+            new_layout[real_qubit] = circuit._layout[faulty_qubits_map[real_qubit]]
+        else:
+            if real_qubit in faulty_qubits:
+                new_layout[real_qubit] = faulty_qreg[faulty_qubit]
+                faulty_qubit += 1
+            else:
+                new_layout[real_qubit] = disconnected_qreg[disconnected_qubit]
+                disconnected_qubit += 1
+    physical_layout_dict = {}
+    for qubit in circuit.qubits:
+        physical_layout_dict[qubit] = faulty_qubits_map_reverse[qubit.index]
+    for qubit in faulty_qreg[:] + disconnected_qreg[:]:
+        physical_layout_dict[qubit] = new_layout[qubit]
+    dag_circuit = circuit_to_dag(circuit)
+    apply_layout_pass = ApplyLayout()
+    apply_layout_pass.property_set['layout'] = Layout(physical_layout_dict)
+    circuit = dag_to_circuit(apply_layout_pass.run(dag_circuit))
+    circuit._layout = new_layout
+    return circuit
+
+
+def _remap_layout_faulty_backend(layout, faulty_qubits_map):
+    if layout is None:
+        return layout
+    new_layout = Layout()
+    for virtual, physical in layout.get_virtual_bits().items():
+        if faulty_qubits_map[physical] is None:
+            raise TranspilerError("The initial_layout parameter refers to faulty"
+                                  " or disconnected qubits")
+        new_layout[virtual] = faulty_qubits_map[physical]
+    return new_layout
```

Comment:
So the thing I don't really like about this is that this is baked into `transpile()` instead of being a pass? It feels to me like this should be a pass that we add to the preset passmanagers during the layout phase. I assume this is just because `CouplingMap` is an `__init__` parameter to passes that use it, but I think this raises a larger question if coupling map isn't a static property for the life of a pass manager anymore then maybe we should be treating it differently in the data flow (like make it part of the property set).

Reply:
you assumption is correct. It use to be part of the design that passes cannot change "static" element like the coupling map. Maybe another thing to revisit.